### PR TITLE
feat: support "any" and "never" within mapped types

### PR DESCRIPTION
**Pull request by sramam**
Example of now supported interface:  [test/valid-data/type-mapped-any-never/main.ts](https://github.com/tufan-io/ts-json-schema-generator/blob/99247bb28d5497c070a8ee25fd16207c2bdf9e6e/test/valid-data/type-mapped-any-never/main.ts#L8) 

This change does change (break?) the [type-conditional-enum](https://github.com/tufan-io/ts-json-schema-generator/blob/99247bb28d5497c070a8ee25fd16207c2bdf9e6e/test/valid-data/type-conditional-enum/main.ts#L8) test, [like so](https://github.com/tufan-io/ts-json-schema-generator/commit/b581cd52d76040631083b68d47188b98429f6b4e#diff-bfab839059303e52b77f8527ffddcc7cacb51cfc4862ce40f6a9d6b9d2c4c070)

I'm not sure on the best path forward. The current implementation supports a wider TS language feature set, but the type-conditional-enum change makes the schema generated in this specific case less elegant.

### Root cause
This seems to stem from a deeper (loose) implementation of how the never keyword is handled.
The root of this is the definition of [Exclude operator](https://github.com/microsoft/TypeScript/blob/d18b728e0b5491662d3513dd346d039a746212f2/src/lib/es5.d.ts#L1486)'s use of never. 

[ConditionalTypeNodeParser.ts](https://github.com/tufan-io/ts-json-schema-generator/blob/eb0676aa291df284d716288b5199d108c10ffb20/src/NodeParser/ConditionalTypeNodeParser.ts#L33) splits the enum up into two - the `trueCheckType` and the `falseCheckType`. With the `trueCheckType` being "excluded", TS implementes it as a "never". The resulting condition is no longer a valid enum, resulting in the new output.

### Positive side-effect(s)
1.  [enables tests to run on Windows](https://github.com/vega/ts-json-schema-generator/compare/master...tufan-io:type-mapped-any-never#diff-9ee9b7ecc4b8f29ba22a091d663854391024be8cb2678c86a58d35008b425717).
2. [schema.json update controlled via env variable](https://github.com/vega/ts-json-schema-generator/compare/master...tufan-io:type-mapped-any-never?expand=1#diff-1a258b987746216dc81c746069ca5944c0f1ec848aa095a8a9f109a976dc2343)

### CODE REVIEW

File: test/utils.ts [hunk 1]
```diff
@@ -47,11 +47,13 @@ export function assertValidSchema(
         const actual: any = JSON.parse(JSON.stringify(schema));
 
         // uncomment to write test files
-        // writeFileSync(
-        //     resolve(`${basePath}/${relativePath}/schema.json`),
-        //     JSON.stringify(schema, null, 4) + "\n",
-        //     "utf8"
-        // );
+        if (process.env.UPDATE_SCHEMA) {
```
Comment (domoritz) [inline test/utils.ts:50]: Awesome. Thank you for this. Can you add a script to the package.json to run the tests with the env var set?

File: test/utils.ts [hunk 2]
```diff
@@ -47,11 +47,13 @@ export function assertValidSchema(
...
+            writeFileSync(
+                resolve(`${basePath}/${relativePath}/schema.json`),
+                JSON.stringify(schema, null, 4) + "\n",
```
Comment (domoritz) [inline test/utils.ts:53]: Let's use `import stringify from "json-stable-stringify";`. It's what we use by default to generate schema files already.

**Review by domoritz — CHANGES_REQUESTED** [review]
* .gitattributes is not needed

The schema for https://github.com/vega/vega-lite has additional properties that should not be there after this pr. 

Please clone the repo, run use yarn to link to the schema generator and run `yarn schema`. Then see the diff.

Screen Shot 2021-02-11 at 13 22 20

Maybe you can send a separate pr for unrelated changes.

File: package.json [hunk 3]
```diff
@@ -84,6 +84,8 @@
     "test": "jest test/ --verbose",
     "test:fast": "FAST_TEST=1 jest test/ --verbose",
     "debug": "node -r ts-node/register --inspect-brk ts-json-schema-generator.ts",
+    "dbg": "node -r ts-node/register --inspect-brk ts-json-schema-generator.ts -p ../template-provider/src/component/template-component/.schema/index.ts -t IConfig",
```
[the remark this answers is not in the document]
Comment (sramam) [reply package.json]: removed. 

> * .gitattributes is not needed

I'll remove it, but I saw a build error - https://github.com/vega/ts-json-schema-generator/runs/1878585120?check_suite_focus=true which prompted me to add it in. 

Not sure what the real problem was, but only experienced it with one file - `test\valid-data\type-aliases-recursive-anonymous\main.ts`. Recreating the file on my (windows) machine solved the problem. 

Would you mind if I added `windows-latest` to the github action? It currently works on windows - would be a shame to let a regression creep back in.

> 
> The schema for https://github.com/vega/vega-lite has additional properties that should not be there after this pr.
> 
> Please clone the repo, run use yarn to link to the schema generator and run `yarn schema`. Then see the diff.
> 
> Screen Shot 2021-02-11 at 13 22 20
> 
> Maybe you can send a separate pr for unrelated changes.

I understand your request. Would you mind accepting this PR as is? 
The "unrelated" parts of this PR were a direct consequence of resolving the underlying issue. 
And I'm way over the time budget to get past this fix.

Some other changes are coming too - which are nicely isolated and will come as separate nicely contained PRs.

[the remark this answers is not in the document]
Comment (sramam) [reply package.json]: removed.

File: test/utils.ts (hunk 1 above)
Comment (sramam) [reply test/utils.ts:50]: done.

File: package.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (domoritz) [reply package.json]: > Would you mind if I added windows-latest to the github action? It currently works on windows - would be a shame to let a regression creep back in.

Go for it. Thanks! 

> I understand your request. Would you mind accepting this PR as is?

Unfortunately, no. I am using this library exclusively for Vega-Lite and therefore cannot accept regressions. I hope you understand this limitation.

**Comment by sramam** [comment]
> 
> The schema for https://github.com/vega/vega-lite has additional properties that should not be there after this pr.
> 
> Please clone the repo, run use yarn to link to the schema generator and run `yarn schema`. Then see the diff.
> 

I can reproduce the problem - but will unfortunately not have the time to understand all that is going on here. 
Is there a smaller example that you can reduce this problem to?

**Comment by domoritz** [comment]
Unfortunately, I won't have the time to reduce the error to a smaller example.

File: package.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (sramam) [reply package.json]: > Unfortunately, I won't have the time to reduce the error to a smaller example.

ouch. I don't know what to do here.

I am in violent agreement with you that regressions are not acceptable.
I'm happy to fix anything I can find and understand. 

Even the additional test case that this PR resolves, was extracted from a much more complex TypeScript project at significant effort to isolate the issue. The net effect however is that I am under intense pressure to move along.

Hopefully that doesn't strike you as being too unreasonable.

[the remark this answers is not in the document]
Comment (sramam) [reply package.json]: I can't even find the "KeepSignal" definition in the repo:
```
/tmp/vega-lite$ pt KeepSignal src/
src//vega.schema.ts
57:export type MapExcludeAndKeepSignalAs<T, E, S extends ExprRef | SignalRef> = {
64:export type MapExcludeValueRefAndReplaceSignalWith<T, S extends ExprRef | SignalRef> = MapExcludeAndKeepSignalAs<
```

[the remark this answers is not in the document]
Comment (domoritz) [reply package.json]: > Hopefully that doesn't strike you as being too unreasonable.

Absolutely not. Developing this library is very tricky and every time I have to make a change, it takes me an hour just to (mentally) page things in again. Debugging large types is tricky and I haven't found a better way than incrementally trying different examples and simplifying them.  

The type is coming from https://github.com/vega/vega/blob/ca27004f8c3ad8aa94f017fb5df6bf90748bda92/packages/vega-typings/types/spec/config.d.ts#L39.

**Comment by sramam** [comment]
@domoritz, I think this fix resolves the issue you raised with vega-lite. 

I don't like it, because it's handled as a post-generation cleanup, and I do not understand the underlying cause.

Now the coverage threshold fails.

I have given it my best shot, but cannot construct a test case that exercises the new code added.

At this point, if you reject this PR, I will have to give up.

**Comment by domoritz** [comment]
Can you rebase this pull request/ merge master?

**Comment by sramam** [comment]
> Can you rebase this pull request/ merge master?

done.

**Comment by domoritz** [comment]
Thank you for your hard work on this pull request!

**Comment by domoritz** [comment]
Unfortunately, this pull request introduced a regression (see Vega-Lite schema below). I will revert it. 

Screen Shot 2021-02-17 at 18 17 58

**Comment by sramam** [comment]
I noticed this- but by design, all empty interfaces are removed. With a default of additionalProps being false, what value does an empty interface provide?

I’m only trying to understand the use of an empty interface.

**Comment by domoritz** [comment]
Unfortunately, the schema now contains `{}`, which seems wrong, no?

**Comment by domoritz** [comment]
I extracted the non-breaking changes so it's easier to see what goes on in this pull request: https://github.com/vega/ts-json-schema-generator/pull/668.

**Comment by sramam** [comment]
> I extracted the non-breaking changes so it's easier to see what goes on in this pull request: #668.

thank you. Sorry you had to do this. 
But now I'm on my best behavior - going forward, you won't have to do it on my account! 

> Unfortunately, the schema now contains {}, which seems wrong, no?

I agree - but in the original (and now current) incarnation, so does `GeoJsonFeature` & `GeoJsonFeatureCollection`.

I'd rather fix these at construction time, but can't seem to figure the minimal examples that reproduce this problem.

Currently, the "empties" removal only removes one level of empty leaves. 
As you point out, that creates new leaves that are now empty.
Perhaps it should be implemented as a recursive call.  

Another idea is to return `undefined` from the NeverTypeFormatter.

I'll try both and report back. 

Sorry again that you are having to jump through hoops to support this change.

**Comment by domoritz** [comment]
Hmm, I guess `GeoJsonFeature` and `GeoJsonFeatureCollection` are any and therefore we get {} (I think {} means any in json schema, not none). 

Thank you again for your help. This program is inherently difficult since we are mapping between different semantics (ts to json). 

When I implemented support for never for the first time, I indeed decided to just return undefined.

No need to apologize. I appreciate pull requests. I apologize for failing to see the regression earlier. I tested the pull request but somehow didn't get the regression in Vega-Lite. I must have forgotten to build.

**Comment by sramam** [comment]
> I think {} means any in json schema, not none

You are right.

Wonder if the right thing to do is the same as undefined: `{ not: {}}` 
Currently, when `NeverTypeNodeParser` returns an undefined when creating the type, it essentially wipes the neverType from the generated schema. While acceptable at one level, it also allows for that property being re-added in the parent scope, depending on how `additionalProperties/additionalItems` is configured. While in general that is to be expected, in this case, the type-system explicitly forbade it, but we lost the information. 

Setting it to `{not: {}}` I think creates the right mapping, even if the schema looks a bit ugly. I'd (ever so slightly) favor `correctness` over `clean`.

Thoughts?

**Comment by domoritz** [comment]
I would prefer to generate clean schema when possible. We should not output not:{} as an option in an enum when we could just omit it for example.

**Comment by sramam** [comment]
@domoritz  I was working under the assumption that post-processing of the schema in the `vega-lite` repo should be considered out of bounds to make this PR acceptable.

Given that you are [already massaging the output](https://github.com/vega/vega-lite/blob/master/scripts/rename-schema.sh) of `ts-json-schema-generator` to fit the needs of `vega-lite`, would you consider doing the same to avoid `vega-lite` regressions from this PR? 

I'd be happy to take a shot at removing the "`empties`" code from here, and add appropriate `KeepSignal` filters to the `rename-schema.sh` file to see if that could be made to work.

Coupled with #672, it'd make this a maintainable solution I imagine. But that decision is beyond my pay-grade. 
Please do let me know.

**Comment by domoritz** [comment]
I think removing empty objects from the output is something we should be able to do here when we output the definitions, no? If not, I would be okay with a final clean up pass (like the one to remove unreachable refs I added a while ago). But post processing should always be a last resort. 

The adjustments we are doing in Vega-Lite are because we want different names in the schema.

**Pull request merged**